### PR TITLE
Added JSON-RPC back-end to Project API and removed deprecated VFS classes

### PR DESCRIPTION
**Pull request by dkulieshov**
Work is still in progress.
Original issue: https://github.com/eclipse/che/issues/6122

Changes to Project Service:
- Added or reworked components according to https://github.com/eclipse/che/issues/6219
- Refactoring and decomposition
- Reworked some places according to new FS API

Changes to Project Manager:
- Refactoring and decomposition
- Reworked package structure
- Split Guice configuration into several classes
- Reworked architecture (where it is possible) to comply DI pattern

Changes to VFS
- Removed all deprecated VFS related classes
- Rearranged code accordingly
- Provided new FS API where needed

Changes to Search
- Reworked index classes that depend on VFS

What is yet to be done:
- [ ] Pass selenium tests
- [x] Rework/add unit tests
- [x] Rework/add java docs

### CODE REVIEW

File: ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/project/ProjectServiceClient.java [hunk 1]
```diff
@@ -160,6 +160,13 @@
   public Promise<Void> importProject(Path path, SourceStorageDto source) {
     String url = getBaseUrl() + IMPORT + encodePath(path);
 
+    if (url.contains("?")) {
+      url += "&clientId=" + appContext.getApplicationId().orElse("");
+    } else {
+
```
Comment (skabashnyuk) [inline ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/project/ProjectServiceClient.java]: Both parts of ```if``` should look the same.

File: ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/project/ProjectServiceClient.java [hunk 2]
```diff
@@ -258,7 +266,15 @@
    */
   public Promise<List<ProjectConfigDto>> createBatchProjects(
       List<NewProjectConfigDto> configurations) {
-    final String url = getBaseUrl() + BATCH_PROJECTS;
+    String url = getBaseUrl() + BATCH_PROJECTS;
+
+    if (url.contains("?")) {
+      url += "&clientId=" + appContext.getApplicationId().orElse("");
+    } else {
+
```
Comment (skabashnyuk) [inline ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/project/ProjectServiceClient.java]: Both parts of ```if``` should look the same.

File: ide/che-core-ide-app/src/test/java/org/eclipse/che/ide/project/ProjectServiceClientTest.java [hunk 3]
```diff
@@ -182,11 +178,11 @@ public void shouldEncodeUrlAndResolveProjectSources() {
 
   @Test
   public void shouldEncodeUrlAndImportProject() {
-    client.importProject(resourcePath, source);
-
-    verify(requestFactory).createPostRequest(any(), eq(source));
-    verify(asyncRequest).header(CONTENT_TYPE, APPLICATION_JSON);
-    verify(asyncRequest).send();
+    //    client.importProject(resourcePath, source);
+    //
```
Comment (skabashnyuk) [inline ide/che-core-ide-app/src/test/java/org/eclipse/che/ide/project/ProjectServiceClientTest.java]: remove if not needed.

File: ide/che-core-ide-app/src/test/java/org/eclipse/che/ide/project/ProjectServiceClientTest.java [hunk 4]
```diff
@@ -211,34 +207,34 @@ public void shouldEncodeUrlAndSearchResourceReferences() {
...
-    verify(asyncRequest).header(ACCEPT, MimeType.APPLICATION_JSON);
-    verify(loaderFactory).newLoader("Creating the batch of projects...");
-    verify(asyncRequest).loader(messageLoader);
-    verify(asyncRequest).send(unmarshallablePrjsConf);
-    verify(unmarshaller).newListUnmarshaller(ProjectConfigDto.class);
-
-    assertEquals(2, prjsArgCaptor.getValue().size());
+    //    List<NewProjectConfigDto> configs = Arrays.asList(prjConfig1, prjConfig2);
+    //
+    //    client.createBatchProjects(configs);
```
Comment (skabashnyuk) [inline ide/che-core-ide-app/src/test/java/org/eclipse/che/ide/project/ProjectServiceClientTest.java]: remove if not needed.

File: plugins/plugin-composer/che-plugin-composer-server/src/main/java/org/eclipse/che/plugin/composer/server/projecttype/ComposerValueProviderFactory.java [hunk 5]
```diff
@@ -61,10 +62,8 @@ protected ComposerValueProvider(FolderEntry projectFolder) {
       }
     }
 
-    private JsonObject readModel(FolderEntry projectFolder) throws ServerException, IOException {
-      FileEntry composerFile = (FileEntry) projectFolder.getChild("composer.json");
-      Reader reader = new BufferedReader(new InputStreamReader(composerFile.getInputStream()));
-      return new Gson().fromJson(reader, JsonObject.class);
+    private JsonObject readModel(Path projectFsPath) throws ServerException, IOException {
+      return new Gson().fromJson(Files.newBufferedReader(projectFsPath), JsonObject.class);
```
Comment (skabashnyuk) [inline plugins/plugin-composer/che-plugin-composer-server/src/main/java/org/eclipse/che/plugin/composer/server/projecttype/ComposerValueProviderFactory.java:74]: Do we need to close reader here ?

File: plugins/plugin-cpp/che-plugin-cpp-lang-server/src/main/java/org/eclipse/che/plugin/cpp/generator/CProjectGenerator.java [hunk 6]
```diff
@@ -10,33 +10,41 @@
...
-    FolderEntry baseFolder = new FolderEntry(vfs.getRoot().createFolder(projectPath.toString()));
-    baseFolder.createFile(
-        FILE_NAME, getClass().getClassLoader().getResourceAsStream("files/default_c_content"));
+      String projectWsPath, Map<String, AttributeValue> attributes, Map<String, String> options)
+      throws ForbiddenException, ConflictException, ServerException, NotFoundException {
+
+    fsManager.createDir(projectWsPath);
+    InputStream inputStream = getClass().getClassLoader().getResourceAsStream(RESOURCE_NAME);
+    String wsPath = resolve(projectWsPath, FILE_NAME);
+    fsManager.createFile(wsPath, inputStream);
```
Comment (skabashnyuk) [inline plugins/plugin-cpp/che-plugin-cpp-lang-server/src/main/java/org/eclipse/che/plugin/cpp/generator/CProjectGenerator.java:47]: do we need to close inputStream here?

File: plugins/plugin-cpp/che-plugin-cpp-lang-server/src/main/java/org/eclipse/che/plugin/cpp/generator/CppProjectGenerator.java [hunk 7]
```diff
@@ -10,33 +10,40 @@
...
-      Path projectPath, Map<String, AttributeValue> attributes, Map<String, String> options)
-      throws ForbiddenException, ConflictException, ServerException {
-    VirtualFileSystem vfs = virtualFileSystemProvider.getVirtualFileSystem();
-    FolderEntry baseFolder = new FolderEntry(vfs.getRoot().createFolder(projectPath.toString()));
-    baseFolder.createFile(
-        FILE_NAME, getClass().getClassLoader().getResourceAsStream("files/default_cpp_content"));
+      String projectWsPath, Map<String, AttributeValue> attributes, Map<String, String> options)
+      throws ForbiddenException, ConflictException, ServerException, NotFoundException {
+    fsManager.createDir(projectWsPath);
+    InputStream inputStream = getClass().getClassLoader().getResourceAsStream(RESOURCE_NAME);
```
Comment (skabashnyuk) [inline plugins/plugin-cpp/che-plugin-cpp-lang-server/src/main/java/org/eclipse/che/plugin/cpp/generator/CppProjectGenerator.java:44]: do we need to close inputStream here?

File: plugins/plugin-java-debugger/che-plugin-java-debugger-server/src/test/java/org/eclipse/che/plugin/jdb/server/util/ProjectApiUtils.java [hunk 8]
```diff
@@ -61,86 +35,88 @@ public static void ensure() throws Exception {
...
+    //    File indexDir = new File("target/fs_index");
+    //    assertTrue(indexDir.mkdirs());
+    //
+    //    Set<PathMatcher> filters = new HashSet<>();
+    //    filters.add(path -> true);
+    //    FSLuceneSearcherProvider sProvider = new FSLuceneSearcherProvider(indexDir, filters);
+    //
+    //    EventService eventService = new EventService();
+    //    LocalVirtualFileSystemProvider vfsProvider =
+    //        new LocalVirtualFileSystemProvider(root, sProvider);
```
Comment (skabashnyuk) [inline plugins/plugin-java-debugger/che-plugin-java-debugger-server/src/test/java/org/eclipse/che/plugin/jdb/server/util/ProjectApiUtils.java]: remove?

File: plugins/plugin-java/che-plugin-java-ext-lang-server/src/main/java/org/eclipse/che/plugin/java/server/rest/JavaFormatterService.java [hunk 9]
```diff
@@ -137,38 +140,28 @@ public void updateProjectFormatter(
...
-    FolderEntry baseFolder = project.getBaseFolder();
     try {
-      VirtualFileEntry cheFolder = baseFolder.getChild(CHE_FOLDER);
-      if (cheFolder == null) {
-        cheFolder = baseFolder.createFolder(CHE_FOLDER);
+      String projectWsPath = absolutize(projectPath);
+      String projectCheFolderWsPath = resolve(projectWsPath, CHE_FOLDER);
+
+      if (!fsManager.existsAsDir(projectCheFolderWsPath)) {
+        fsManager.createDir(projectCheFolderWsPath);
```
Comment (skabashnyuk) [inline plugins/plugin-java/che-plugin-java-ext-lang-server/src/main/java/org/eclipse/che/plugin/java/server/rest/JavaFormatterService.java:148]: do we need to check if a directory has been created here?

File: plugins/plugin-java/che-plugin-java-ext-lang-server/src/main/java/org/eclipse/che/plugin/java/server/rest/JavaFormatterService.java [hunk 10]
```diff
@@ -137,38 +140,28 @@ public void updateProjectFormatter(
...
       }
-      FolderEntry cheFolderEntry = (FolderEntry) cheFolder;
-      VirtualFileEntry formatter = cheFolderEntry.getChild(CHE_FORMATTER_XML);
 
-      if (formatter == null) {
-        cheFolderEntry.createFile(CHE_FORMATTER_XML, content.getBytes());
+      String cheFormatterWsPath = resolve(projectCheFolderWsPath, CHE_FORMATTER_XML);
+
+      if (!fsManager.existsAsFile(cheFormatterWsPath)) {
+        fsManager.createFile(cheFormatterWsPath, content);
```
Comment (skabashnyuk) [inline plugins/plugin-java/che-plugin-java-ext-lang-server/src/main/java/org/eclipse/che/plugin/java/server/rest/JavaFormatterService.java:154]: do we need to check if a file has been created here?

File: plugins/plugin-composer/che-plugin-composer-server/src/main/java/org/eclipse/che/plugin/composer/server/projecttype/ComposerValueProviderFactory.java (hunk 5 above)
Comment (dkulieshov) [reply plugins/plugin-composer/che-plugin-composer-server/src/main/java/org/eclipse/che/plugin/composer/server/projecttype/ComposerValueProviderFactory.java:74]: I'm not familiar with the logic, I did just source code adaptation to a new API, so I don't really know if we need to close reader here. If it matters for you, I guess you can ask an author.

**Comment by dkulieshov** [comment]
ci-test

File: ide/che-core-ide-app/src/test/java/org/eclipse/che/ide/project/ProjectServiceClientTest.java (hunk 4 above)
Comment (dkulieshov) [reply ide/che-core-ide-app/src/test/java/org/eclipse/che/ide/project/ProjectServiceClientTest.java]: This test is already reworked be one of my previous commits, in general I'm not sure if it is a good idea to review the PR that is in a work-in-progress state, I mean you will have many of your remarks outdated and useless. Please watch "status" label to be aware of current PR status not to waste your time.

File: plugins/plugin-cpp/che-plugin-cpp-lang-server/src/main/java/org/eclipse/che/plugin/cpp/generator/CProjectGenerator.java (hunk 6 above)
Comment (dkulieshov) [reply plugins/plugin-cpp/che-plugin-cpp-lang-server/src/main/java/org/eclipse/che/plugin/cpp/generator/CProjectGenerator.java:47]: Again, I'm not sure if we need to close the stream here, probably we do, however this is out of scope of this issue. The PR already has like three hundreds of file changed, honestly, I see those and some other errors, but I haven't the slightest desire to mix those fixes with this huge PR. I'm sure that there will be enough PR-related bugs and errors.

If you are really concerned about such kind of errors (and I sincerely share your concern) I guess you can create a dedicated issue so we could later get back to it and fix all those bugs.

**Comment by dkulieshov** [comment]
ci-test

**Review by skabashnyuk — APPROVED** [review]

**Comment by dkulieshov** [comment]
ci-test

File: plugins/plugin-cpp/che-plugin-cpp-lang-server/src/main/java/org/eclipse/che/plugin/cpp/generator/CProjectGenerator.java (hunk 6 above)
Comment (dkulieshov) [reply plugins/plugin-cpp/che-plugin-cpp-lang-server/src/main/java/org/eclipse/che/plugin/cpp/generator/CProjectGenerator.java:47]: Created an issue: https://github.com/eclipse/che/issues/6841.

File: ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/project/ProjectServiceClient.java (hunk 1 above)
Comment (dkulieshov) [reply ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/project/ProjectServiceClient.java]: Improved by this [commit](https://github.com/eclipse/che/pull/6726/commits/013c55ec39634563b22afe00e1d5465de5c21aba)

**Comment by dkulieshov** [comment]
ci-test

**Comment by dkulieshov** [comment]
ci-test

**Comment by dkulieshov** [comment]
ci-test

**Comment by dkulieshov** [comment]
ci-test

**Comment by dkulieshov** [comment]
ci-test

**Comment by dkulieshov** [comment]
ci-test

**Comment by musienko-maxim** [comment]
ci-test

**Comment by dkulieshov** [comment]
ci-test

**Comment by dkulieshov** [comment]
ci-test

**Comment by dkulieshov** [comment]
ci-test

**Comment by dkulieshov** [comment]
ci-test

**Comment by dkulieshov** [comment]
ci-test

**Comment by dkulieshov** [comment]
ci-test

**Comment by dkulieshov** [comment]
ci-test

**Comment by dkulieshov** [comment]
ci-test

**Review by vparfonov — APPROVED** [review]

**Comment by dkulieshov** [comment]
ci-test

**Comment by dkulieshov** [comment]
ci-test

**Comment by benoitf** [comment]
it is normal this PR has been closed ? looks like it was a work in progress since one month ?

**Comment by dkulieshov** [comment]
Created a new PR instead of this one https://github.com/eclipse/che/pull/7380

**Pull request closed without merging**